### ISSUE TITLE
[RFC] Modeling of Interaction with other dialects

### ISSUE THREAD

**Issue by martin-luecke** [rise-lang/mlir#5]
Currently we use `rise.unwrap` and `rise.wrap` to warp and unwrap scalar types for usage with other dialects. This looks like this:
```
%add = rise.lambda (%summand0, %summand1) : !rise.fun<scalar<f32> -> fun<scalar<f32> ->
   scalar<f32>>> {
    %a = rise.unwrap %summand0
    %b = rise.unwrap %summand1
    %res = addf %a, %b : f32
    %result = rise.wrap %res
    rise.return %result : !rise.scalar<f32>
}
```
Currently our lowering assumes that all operations from other dialects are enclosed by the `rise.unwrap` and the `rise.wrap`. We could also model this implicit assumption explicitly using a region:
```
%add = rise.lambda (%summand0, %summand1) : !rise.fun<scalar<f32> -> fun<scalar<f32> ->
   scalar<f32>>> {
    %result = rise.unwrap(%a = %summand0, %b = %summand1) : scalar<f32> {
        %res = addf %a, %b : f32
        rise.return %res
    }
    rise.return %result : !rise.scalar<f32>
}
```
The new `rise.unwrap` only exposes the wrapped type inside its region and wraps the returned result of its region for further use within the rise dialect. I expect this approach also simplifies rewriting programs containing lots of operations from other dialects, as the rise.unwrap operation represents all contained foreign operations.

**Comment by michel-steuwer** [comment]
That looks like a meaningful design to me, allowing to reduce the number of `rise.warp/unwrap` necessary.

Can we think of a different name for the proposed `rise.unwrap`? Maybe something like `rise.embed` ... suggestions welcome.

We should avoid any `rise` specific stuff inside the wrapped regions (so no `rise.return` but a standard `return`).

I could imagine an additional variation of this idea, a `rise.call` for calling a MLIR `func` while performing the type (un)wrapping on the arguments and return value.

**Comment by martin-luecke** [comment]
I think we could also (at least while pretty printing) avoid the renaming in the unwrapping. So this:
```
%result = rise.embed(%a = %summand0, %b = %summand1) : scalar<f32> {
    %res = addf %a, %b : f32
    return %res
}
```
Would turn into this:
```
%result = rise.embed(%summand0, %summand1) : scalar<f32> {
    %res = addf %summand0, %summand1 : f32
    return %res
}
```
The arguments to the unwrap will be unwrapped inside the region and shadow the outside values which have the same name.

**Comment by martin-luecke** [comment]
Implemented design + lowering in 075fdf2687d080547c2ca071561dd382ee495171
